wkcuber/recubing.py

```diff
@@ -0,0 +1,138 @@
+    )
+
+    parser.add_argument(
+        "--dtype",
+        "-d",
+        help="Target datatype (e.g. uint8, uint16, uint32)",
+        default="uint8",
+    )
+
+    parser.add_argument(
+        "--wkw_file_len", help="Target file length (default 32)", type=int, default=32
+    )
+
+    parser.add_argument(
+        "--no_compression",
+        help="Use compression, default false",
+        type=bool,
+        default=False,
+    )
+
+    add_verbose_flag(parser)
+    add_distribution_flags(parser)
+
+    return parser
+
+
+def next_lower_divisible_by(number, divisor) -> int:
+    remainder = number % divisor
+    return number - remainder
+
+
+def next_higher_divisible_by(number, divisor) -> int:
+    remainder = number % divisor
+    return number - remainder + divisor
+
+
+def recube(
+    source_path, target_path, layer_name, dtype, wkw_file_len=32, compression=True
+):
+    if compression:
+        block_type = wkw.Header.BLOCK_TYPE_LZ4
+    else:
+        block_type = wkw.Header.BLOCK_TYPE_RAW
+
+    target_wkw_header = wkw.Header(
+        np.dtype(dtype), file_len=wkw_file_len, block_type=block_type
+    )
+    target_wkw_info = WkwDatasetInfo(target_path, layer_name, 1, target_wkw_header)
+    source_wkw_header = wkw.Header(np.dtype(dtype))
+    source_wkw_info = WkwDatasetInfo(source_path, layer_name, 1, source_wkw_header)
+
+    ensure_wkw(target_wkw_info)
+
+    bounding_box = infer_bounding_box(source_wkw_info)
+    bottom_right = [
+        coord + size for coord, size in zip(bounding_box[0], bounding_box[1])
+    ]
+
+    wkw_cube_size = wkw_file_len * BLOCK_LEN
+
+    outer_bounding_box_tl = list(
+        map(lambda lx: next_lower_divisible_by(lx, wkw_cube_size), bounding_box[0])
+    )
+    outer_bounding_box_br = list(
+        map(lambda lx: next_higher_divisible_by(lx, wkw_cube_size), bottom_right)
+    )
+    outer_bounding_box_size = [
+        outer_bounding_box_br[0] - outer_bounding_box_tl[0],
+        outer_bounding_box_br[1] - outer_bounding_box_tl[1],
+        outer_bounding_box_br[2] - outer_bounding_box_tl[2],
+    ]
+
+    with open_wkw(source_wkw_info) as source_wkw_dataset:
+        for x in range(0, outer_bounding_box_size[0], wkw_cube_size):
+            for y in range(0, outer_bounding_box_size[1], wkw_cube_size):
+                for z in range(0, outer_bounding_box_size[2], wkw_cube_size):
+                    logging.info("Reading at offset {}, {}, {}".format(x, y, z))
+
+                    top_left = [
+                        outer_bounding_box_tl[0] + x,
+                        outer_bounding_box_tl[1] + y,
+                        outer_bounding_box_tl[2] + z,
+                    ]
+
+                    logging.info("Writing at {}".format(top_left))
+
+                    data_cube = source_wkw_dataset.read(
+                        top_left, (wkw_cube_size, wkw_cube_size, wkw_cube_size)
+                    )
+                    target_wkw_dataset = open_wkw(target_wkw_info)
```

Comment:
This `open_wkw` call should be below line 106 in a `with` context (similar to source_wkw_dataset). Otherwise, you open the dataset for each cube (and it's not closed).